Issue 694: Upgrade/refactoring for U and V write-out sub for FV3REG GSI failure …

TingLei-daprediction opened the pull request:
**DUE DATE for merger of this PR into `develop` is 3/27/2024 (six weeks after PR creation).**

Resolves #693 (Thanks to @edwardhartnett 's suggestions)
Resolves # 694 ( this PR is not able to provide a stable solution, more details would be given on the issue page) 
Resolves # 697:  With larger requested memory for each mpi task, it still showed, for some time, the differences in the analysis files between loproc vs hiproc for the control runs on hercules. whether integrating this  with the refactored IO part would provide a stable solution remains to be seen. 

This PR  resolved the newly emerged issue with IO of netcdf files in the continuous storage, with upgraded  FV3REG IO for the cold start options. (Co author Ming Hu @hu5970 ) 
This PR is being worked on in collaboration with Pete Johnson through RDHPCS help desk,  @RussTreadon-NOAA @DavidHuber-NOAA and thanks to help  from @ed Raghue Reddy through RDHPCS help desk.

TingLei-daprediction committed:
Upgrade/refactoring for U and V write-out sub for FV3REG GSI failure on hercules: GSI ISSUE(#694):

TingLei-daprediction committed:
Upgrade/refactoring for U and V write-out sub for FV3REG GSI failure on hercules: GSI ISSUE(#694):

RussTreadon-NOAA commented:
@TingLei-NOAA , please update [`TingLei-daprediction:feature/fv3reg_parallel_io_upgrade`](https://github.com/TingLei-daprediction/GSI/tree/feature/fv3reg_parallel_io_upgrade) to the current head of NOAA-EMC/GSI [`develop`](https://github.com/NOAA-EMC/GSI/tree/develop). PR #684 is not in your branch.

TingLei-NOAA commented:
@RussTreadon-NOAA  Sure. I am verifying [TingLei-daprediction:feature/fv3reg_parallel_io_upgrade](https://github.com/TingLei-daprediction/GSI/tree/feature/fv3reg_parallel_io_upgrade)  with the current ECM GSI .

RussTreadon-NOAA commented:
PR #684 was merged into NOAA-EMC/GSI `develop` yesterday afternoon (1:07 pm EST, 2/14/2024).  Your branch is one commit behind the current EMC GSI `develop`.

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'develop' into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'develop' into feature/fv3reg_parallel_io_upgrade

Commit:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

hu5970 commented:
@TingLei-NOAA The  RRFS does not use the "fv3_io_layout_y > 1" option anymore.

TingLei-daprediction committed:
update feature/fv3reg_parallel_io_upgrade with changes to writeuv_v1 for cold start files and add nf90_collective mode for U and V following the suggestion from P. Johson (through Hercules helpdesk)

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-NOAA commented:
An update:  Thanks to discussions/collaborations with Peter Johnson at Hercules help desk , @RussTreadon-NOAA, @DavidHuber-NOAA , @edwardhartnett and other colleagues, it is found : 1), I agree with Peter Johnson's speculation that the culprit is with the mpi lib on herculues, while further experiments could be run like using different mpi lib to confirm this if needed; 2)  the current PR https://github.com/NOAA-EMC/GSI/pull/698 seems to be a solution/work-around when it has been running successfully in my runs of 400 plus .  It should be noted, without the nf90_collective mode added as Peter Johnson proposed, GSI would fail once in 50 plus runs.   The issue was found and resolved for the warm-restart case, but the similar change has also been added to the write-out subroutine for the cold start cases:  gsi_fv3ncdf_writeuv_v1.  The verification of this part is to be updated later.

RussTreadon-NOAA commented:
**Hercules `hafs_3denvar_hybens`**

Build `develop` at 74ac5942 and `TingLei-daprediction:feature/fv3reg_parallel_io_upgrade` at e0dc8d05 on Hercules.


The `hafs_3denvar_hybens` ctest fails due to non-reproducible `fv3_tracer` files.  The _hiproc_contrl_, _loproc_contrl_, and _loproc_updat_ `fv3_tracer` are identical.  The _hiproc_updat_ differs for variable `sphum`
```
xaxis_1 min/max 1=1.0,720.0 min/max 2=1.0,720.0 max abs diff=0.0000000000
yaxis_1 min/max 1=1.0,540.0 min/max 2=1.0,540.0 max abs diff=0.0000000000
zaxis_1 min/max 1=1.0,65.0 min/max 2=1.0,65.0 max abs diff=0.0000000000
Time min/max 1=1.0,1.0 min/max 2=1.0,1.0 max abs diff=0.0000000000
sphum min/max 1=0.0,0.022951916 min/max 2=0.0,0.022951916 max abs diff=0.0000935391
liq_wat min/max 1=-1.27048754e-20,0.001674175 min/max 2=-1.27048754e-20,0.001674175 max abs diff=0.0000000000
rainwat min/max 1=-6.4782625e-19,0.0052170595 min/max 2=-6.4782625e-19,0.0052170595 max abs diff=0.0000000000
ice_wat min/max 1=-9.211089e-21,0.0016910682 min/max 2=-9.211089e-21,0.0016910682 max abs diff=0.0000000000
snowwat min/max 1=-4.735845e-20,0.0022513575 min/max 2=-4.735845e-20,0.0022513575 max abs diff=0.0000000000
graupel min/max 1=-8.9380465e-20,0.01118745 min/max 2=-8.9380465e-20,0.01118745 max abs diff=0.0000000000
o3mr min/max 1=1.4650413e-08,1.6894655e-05 min/max 2=1.4650413e-08,1.6894655e-05 max abs diff=0.0000000000
sgs_tke min/max 1=-1.38765404e-17,37.100296 min/max 2=-1.38765404e-17,37.100296 max abs diff=0.0000000000
cld_amt min/max 1=0.0,1.0 min/max 2=0.0,1.0 max abs diff=0.0000000000
```

TingLei-NOAA commented:
@RussTreadon-NOAA   Yes this PR itself couldn't solely resolve the issue https://github.com/NOAA-EMC/GSI/issues/697

RussTreadon-NOAA commented:
Oh, I see.  I read your comment

> 2) the current PR https://github.com/NOAA-EMC/GSI/pull/698 seems to be a solution/work-around when it has been running successfully in my runs of 400 plus .

and assumed you had a fix.

TingLei-NOAA commented:
@RussTreadon-NOAA  I meant this PR is the fix/work-around for issue 694 (Parallel netCDF I/O failures on Hercules with I_MPI_EXTRA_FILESYSTEM=1).

RussTreadon-NOAA commented:
Thank for the clarification.  `ush/sub_hercules` in `TingLei-daprediction:feature/fv3reg_parallel_io_upgrade` still includes
```
echo "module load gsi_hercules.intel" >> $cfile
#TODO reenable I_MPI_EXTRA_FILESYSTEM once regional ctests can properly handle parallel I/O on Hercules
echo "unset I_MPI_EXTRA_FILESYSTEM"         >> $cfile
echo ""                                                    >> $cfile
```
Bottom line:  we still have non-reproducible `hafs_3denvar_hybens` results on Hercules.

TingLei-NOAA commented:
@RussTreadon-NOAA Thanks. I will clean that part.

TingLei-NOAA commented:
As mentioned earlier,  the similar refactoring and changes have been done with the write-out of winds for the cold start option.  @ShunLiu-NOAA and @hu5970 recently found, with newer netcdf lib, when the cold start input files are in continuous storage, the  GSI would become idle.   Further investigation confirmed it become idle for those reading mpi processes become idle.  Hence, it is believed that changes as done in https://github.com/NOAA-EMC/GSI/pull/571/  are needed for the reading subroutine for the cold start files and will be included in https://github.com/NOAA-EMC/GSI/pull/698 corresponding to this issue.

RussTreadon-NOAA commented:
**Hercules test**

Install `TingLei-daprediction:feature/fv3reg_parallel_io_upgrade` at e0dc8d05.  Run `hafs_3denvar_hybens` and `hafs_4denvar_glbens`.   Test `hafs_3denvar_hybens` passed.  Test `hafs_4denvar_glbens` failed.

`hafs_4denvar_glbens` failed because output file `fv3_tracer` is not identical between the _loproc_updat_ and _hiproc_updat_ runs.  Differences are found in the `sphum` field
```
xaxis_1 min/max 1=1.0,720.0 min/max 2=1.0,720.0 max abs diff=0.0000000000
yaxis_1 min/max 1=1.0,540.0 min/max 2=1.0,540.0 max abs diff=0.0000000000
zaxis_1 min/max 1=1.0,65.0 min/max 2=1.0,65.0 max abs diff=0.0000000000
Time min/max 1=1.0,1.0 min/max 2=1.0,1.0 max abs diff=0.0000000000
sphum min/max 1=0.0,0.022629073 min/max 2=0.0,0.022629073 max abs diff=0.0000017433
liq_wat min/max 1=-1.27048754e-20,0.001674175 min/max 2=-1.27048754e-20,0.001674175 max abs diff=0.0000000000
rainwat min/max 1=-6.4782625e-19,0.0052170595 min/max 2=-6.4782625e-19,0.0052170595 max abs diff=0.0000000000
ice_wat min/max 1=-9.211089e-21,0.0016910682 min/max 2=-9.211089e-21,0.0016910682 max abs diff=0.0000000000
snowwat min/max 1=-4.735845e-20,0.0022513575 min/max 2=-4.735845e-20,0.0022513575 max abs diff=0.0000000000
graupel min/max 1=-8.9380465e-20,0.01118745 min/max 2=-8.9380465e-20,0.01118745 max abs diff=0.0000000000
o3mr min/max 1=1.4650413e-08,1.6894655e-05 min/max 2=1.4650413e-08,1.6894655e-05 max abs diff=0.0000000000
sgs_tke min/max 1=-1.38765404e-17,37.100296 min/max 2=-1.38765404e-17,37.100296 max abs diff=0.0000000000
cld_amt min/max 1=0.0,1.0 min/max 2=0.0,1.0 max abs diff=0.0000000000
```
The above ctests were run in the `/work` fileset.  

The hafs tests were also run in `/work2`.  `hafs_3denvar_hybens` generated identical output files for _loproc_updat_ and _hiproc_updat_.   Output file `fv3_tracer` differs between the _loproc_ and _hiproc_ runs `hafs_4denvar_glbens`.  The difference is in `o3mr`
```
xaxis_1 min/max 1=1.0,720.0 min/max 2=1.0,720.0 max abs diff=0.0000000000
yaxis_1 min/max 1=1.0,540.0 min/max 2=1.0,540.0 max abs diff=0.0000000000
zaxis_1 min/max 1=1.0,65.0 min/max 2=1.0,65.0 max abs diff=0.0000000000
Time min/max 1=1.0,1.0 min/max 2=1.0,1.0 max abs diff=0.0000000000
sphum min/max 1=0.0,0.022629073 min/max 2=0.0,0.022629073 max abs diff=0.0000000000
liq_wat min/max 1=-1.27048754e-20,0.001674175 min/max 2=-1.27048754e-20,0.001674175 max abs diff=0.0000000000
rainwat min/max 1=-6.4782625e-19,0.0052170595 min/max 2=-6.4782625e-19,0.0052170595 max abs diff=0.0000000000
ice_wat min/max 1=-9.211089e-21,0.0016910682 min/max 2=-9.211089e-21,0.0016910682 max abs diff=0.0000000000
snowwat min/max 1=-4.735845e-20,0.0022513575 min/max 2=-4.735845e-20,0.0022513575 max abs diff=0.0000000000
graupel min/max 1=-8.9380465e-20,0.01118745 min/max 2=-8.9380465e-20,0.01118745 max abs diff=0.0000000000
o3mr min/max 1=1.4650413e-08,1.6894655e-05 min/max 2=1.4650413e-08,1.6894655e-05 max abs diff=0.0000000318
sgs_tke min/max 1=-1.38765404e-17,37.100296 min/max 2=-1.38765404e-17,37.100296 max abs diff=0.0000000000
cld_amt min/max 1=0.0,1.0 min/max 2=0.0,1.0 max abs diff=0.0000000000
```

How does program execution differ between `hafs_3denvar_hybens` and `hafs_4denvar_glbens` with regards to the writing of output file `fv3_tracer`?

TingLei-NOAA commented:
@RussTreadon-NOAA  Thanks. for the hafs issue on the differences between loproc and hiproc, @yonghuiweng had some updates on https://github.com/NOAA-EMC/GSI/issues/697

RussTreadon-NOAA commented:
OK, so what's the path forward to get the hafs ctests consistently passing on Hercules?  Will updates be committed to `TingLei-daprediction:feature/fv3reg_parallel_io_upgrade` in the near future?

TingLei-NOAA commented:
@RussTreadon-NOAA    There will be a google meeting for that issue among Yonghui, Bin and me to get some clarifications on current findings and see how to proceed from our point of view . You will definitely be updated on it and all of us will see how to proceed. Let me know if you have any suggestions for being now.

RussTreadon-NOAA commented:
Sounds good.  We want to get to the bottom of this sooner rather than later.  

This is an odd problem.  The hafs tests pass on WCOSS2, Hera, and Orion.  Orion and Hercules use the same filesets.  My concern is that the Hercules issue is somehow related to Rocky-8, module versions, and/or the installation of supporting libraries on Hercules.  Will we see similar problems when Hera and Orion are updated to Rocky-8 in April?

TingLei-NOAA commented:
Sure  when newer versions of software are used on hercules.  We need to,at least, make sure if the issue is related to hercules system(hardware ) or generically related to the newer versions of software, which would be encountered in the future in other machines.

TingLei-daprediction committed:
Now,for fv3reg cold start files, only mpi processes acctually particpating in IO would be
grouped to create a sub communicator,which would be used in nf90_open command.
which had been done with warm restart files.
Co-authored by: Ming Hu at Ming.Hu@noaa.gov

TingLei-daprediction committed:
Co-authored by: Ming Hu at Ming.Hu@noaa.gov
Now,for fv3reg cold start files, only mpi processes acctually particpating in IO would be
grouped to create a sub communicator,which would be used in nf90_open command.
which had been done with warm restart files.

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Sleaning of print-out lines for debugging

TingLei-daprediction committed:
Sleaning of print-out lines for debugging

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

Commit:
Merge branch 'develop' into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
modification in regress scripts for setup for fv3reg GSI

TingLei-daprediction committed:
modification in regress scripts for setup for fv3reg GSI

TingLei-daprediction commented:
The current PR passed regression tests on hera and wcoss2.  On wcoss2, it also gave identifcal results for a rrfs case compared with the current EMC GSI. 
The regression tests on Hercules is not finished yet.  But it is believed the current PR could move forward for the reviewing process. @ShunLiu-NOAA  Thanks to your comments and suggestions,  now the PR could go ahead.  Thanks.

ShunLiu-NOAA commented:
@JingCheng-NOAA and @guoqing-noaa Could you please review this PR? The changes could have impact on HAFS DA and RTMA.

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
tunning of mpi task layout in regression tests on Hercules

TingLei-daprediction committed:
tunning of mpi task layout in regression tests on Hercules

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
adding nf90_independent for non-wind variables IO in Fv3reg GSI

TingLei-daprediction committed:
tunning of mpi task layout in regression tests on Hercules

DavidHuber-NOAA commented:
@TingLei-NOAA  I recall now that Hercules does not submit exclusive jobs by default.  That is, multiple jobs can be sent to the same node, so I think the requested memory for a job is proportional to the number of cores requested.  Perhaps the `--mem` or `--mem-per-cpu` parameters need to be set when evoking `sbatch` to ensure enough memory is available?

RussTreadon-NOAA commented:
@DavidHuber-NOAA , you are correct.  We should specify `--mem=0` to reserve all memory on the node for the job.  @TingLei-NOAA , check [slurm sbatch documentation](https://slurm.schedmd.com/sbatch.html) for the `--exclusive` and `--mem` for details

RussTreadon-NOAA commented on regression/regression_driver.sh:
```diff
@@ -36,10 +36,12 @@ for jn in `seq ${RSTART} ${REND}`; do
       export scripts=${scripts_updat:-$scripts}
       export fixgsi=${fixgsi_updat:-$fixgsi}
       export modulefiles=${modulefiles_updat:-$modulefiles}
+      export ush=${ush_update:-$ush}
```
What is the reason for adding `ush` here and below?

TingLei-NOAA replied:
@RussTreadon-NOAA  When ush_cntrl is needed and added,  I "feel"  ush_update is needed to make this extra flexibility/controllability  complete as setup for  for modulefiles_control vs modulefiles_update.

RussTreadon-NOAA commented on regression/regression_param.sh:
```diff
@@ -87,7 +87,8 @@ case $regtest in
 
         if [[ "$machine" = "Hera" ]]; then
-           topts[1]="0:15:00" ; popts[1]="20/1/"  ; ropts[1]="/1"
+#cltorg           topts[1]="0:15:00" ; popts[1]="20/1/"  ; ropts[1]="/1"
```
No need to retain commented out lines.  Version control tracks changes.  `git diff` of various hashes will show exactly how a file has changed.

RussTreadon-NOAA commented on ush/sub_hera:
```diff
@@ -120,10 +120,10 @@ echo "#SBATCH --output=$output"
 echo "#SBATCH --qos=$queue"                                                        >> $cfile
 echo "#SBATCH --time=$timew"                                                       >> $cfile
-echo "#SBATCH --nodes=$nodes --ntasks-per-node=$procs --cpus-per-task=$threads"    >> $cfile
+echo "#SBATCH --nodes=$nodes --ntasks-per-node=$procs --cpus-per-task=$threads --exclusive"    >> $cfile
```
If you want to reserve all memory on a node use `--mem=0`.

RussTreadon-NOAA commented on ush/sub_hera:
```diff
@@ -120,10 +120,10 @@ echo "#SBATCH --output=$output"
 #echo "#SBATCH -j oe"                                       >> $cfile
 echo "#SBATCH --account=$accnt"                                                    >> $cfile
-echo "#SBATCH --mem=0"                                                             >> $cfile
+#cltorg echo "#SBATCH --mem=0"                                                             >> $cfile
```
Remove commented out line.

RussTreadon-NOAA commented on ush/sub_hera:
```diff
@@ -143,7 +143,8 @@ echo "module list"                                         >> $cfile
 
 cat $exec >> $cfile
-
+echo 'thinkdeb cat cfile'
```
Are these lines for debugging?  If so, remove before PR is merged into `develop`.

TingLei-NOAA commented:
@DavidHuber-NOAA @RussTreadon-NOAA  Thanks for comments on memory setup, especially on the --exclusive not working, which I added and removed --mem=0.   Because, from my digging, --mem=0 's impacts are machine dependent.   I would seek confirmation/clariciation from Hercluse  help desk.

TingLei-NOAA commented:
An update :  Current PR also passes all regression tests on Hercules ( in two regress test runs).  But, the ppn for hafs_3denvar_*hproc is set to 2.  It looks a little suspicious.   Thanks to @DavidHuber-NOAA and @RussTreadon-NOAA reminding on the "exclusive memory usage " set up on herculse, I will try to increase ppn after a correct setup for mem exclusive usage.  (To run regession tests for this PR, ush_cntrl as a env variable should be defined to point to the ush dir of the control GSI branch for different sub_hercules in ush dir are used for control and update runs).

TingLei-NOAA commented:
An summary on memory setup for slurm jobs on rdhpcs that I learned from Raghue Reddy through hercules help desk:
1) don't use "sbatch --cpus-per-task=N",   which will make on N cpu share of memory of a node to be used by the job.
related doc from Raghue:
"
https://rdhpcs-com:mon-docs.rdhpcs.noaa.gov/wiki/index.php?title=Recent_User_Facing_Changes#Feb_10,_2022:_slurm-21.08.5_update_causing_Out_of_Memory_(OOM)_errors

https://rdhpcs-common-docs.rdhpcs.noaa.gov/wiki/index.php?title=Running_and_Monitoring_Jobs#Setting_the_--cpus-per-task_parameter"
2.  use --mem=0  in the srun command line. 
especially when --cpu-per-task is used in the command line for multi threads jobs. 

I have made changes to the gsi regression tests on hercules and will see if it could avoid the large numbers of nodes used in the current PR.

TingLei-daprediction committed:
corrected use of '--mem=0, --recursive' in sbatch job on Herclues

TingLei-daprediction committed:
tuning of mpi task layout

TingLei-daprediction committed:
tuning of sbatch parameters for fv3reg tests on hera

ShunLiu-NOAA commented:
@guoqing-noaa and @JingCheng-NOAA Could you please review this PR?

TingLei-daprediction committed:
tuning of slurm paramters for  FV3REG GSI regression tests on orion

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
revert to a larger requested nodes number for hafs_4denvar_glbens_hiproc_updat

TingLei-daprediction committed:
tuning slurm parameters for hafs tests on orion

ShunLiu-NOAA commented:
@JingCheng-NOAA and @guoqing-noaa Could you please let me know whether you are available to review this PR?

JingCheng-NOAA commented:
> @JingCheng-NOAA and @guoqing-noaa Could you please let me know whether you are available to review this PR?

I've been doing the regression test on Hercules since Friday. But so far, it seems like something not setting up correctly. I am double checking it to see which part went wrong.
Test project /work2/noaa/hwrf/save/jcheng/GSI/GSItest/build
    Start 1: global_4denvar
1/7 Test #1: global_4denvar ...................   Passed  1326.62 sec
    Start 2: rtma
2/7 Test #2: rtma .............................   Passed  965.50 sec
    Start 3: rrfs_3denvar_glbens
3/7 Test #3: rrfs_3denvar_glbens ..............***Failed  425.44 sec
    Start 4: netcdf_fv3_regional
4/7 Test #4: netcdf_fv3_regional ..............***Timeout 86400.06 sec
    Start 5: hafs_4denvar_glbens
5/7 Test #5: hafs_4denvar_glbens ..............***Timeout 86400.04 sec
    Start 6: hafs_3denvar_hybens

ShunLiu-NOAA commented:
Thank you @JingCheng-NOAA.

TingLei-daprediction commented:
Jing
Did you do this as described in the previous comments “
To run regession tests for this PR, ush_cntrl as a env variable should be
defined to point to the ush dir of the control GSI branch for different
sub_hercules in ush dir are used for control and update runs).”

Regard
Ting

On Mon, Mar 4, 2024 at 7:41 AM JingCheng-NOAA ***@***.***>
wrote:

> @JingCheng-NOAA <https://github.com/JingCheng-NOAA> and @guoqing-noaa
> <https://github.com/guoqing-noaa> Could you please let me know whether
> you are available to review this PR?
>
> I've been doing the regression test on Hercules since Friday. But so far,
> it seems like something not setting up correctly. I am double checking it
> to see which part went wrong.
> Test project /work2/noaa/hwrf/save/jcheng/GSI/GSItest/build
> Start 1: global_4denvar
> 1/7 Test #1 <https://github.com/NOAA-EMC/GSI/pull/1>: global_4denvar
> ................... Passed 1326.62 sec
> Start 2: rtma
> 2/7 Test #2 <https://github.com/NOAA-EMC/GSI/issues/2>: rtma
> ............................. Passed 965.50 sec
> Start 3: rrfs_3denvar_glbens
> 3/7 Test #3 <https://github.com/NOAA-EMC/GSI/issues/3>:
> rrfs_3denvar_glbens ..............***Failed 425.44 sec
> Start 4: netcdf_fv3_regional
> 4/7 Test #4 <https://github.com/NOAA-EMC/GSI/pull/4>: netcdf_fv3_regional
> ..............***Timeout 86400.06 sec
> Start 5: hafs_4denvar_glbens
> 5/7 Test #5 <https://github.com/NOAA-EMC/GSI/pull/5>: hafs_4denvar_glbens
> ..............***Timeout 86400.04 sec
> Start 6: hafs_3denvar_hybens
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/698#issuecomment-1976497248>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADSYWF35GEJNWKKVDH6NWILYWRTWRAVCNFSM6AAAAABDJG6Q3GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZWGQ4TOMRUHA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

JingCheng-NOAA commented:
> Jing Did you do this as described in the previous comments “ To run regession tests for this PR, ush_cntrl as a env variable should be defined to point to the ush dir of the control GSI branch for different sub_hercules in ush dir are used for control and update runs).” Regard Ting
> […](#)
> On Mon, Mar 4, 2024 at 7:41 AM JingCheng-NOAA ***@***.***> wrote: @JingCheng-NOAA <https://github.com/JingCheng-NOAA> and @guoqing-noaa <https://github.com/guoqing-noaa> Could you please let me know whether you are available to review this PR? I've been doing the regression test on Hercules since Friday. But so far, it seems like something not setting up correctly. I am double checking it to see which part went wrong. Test project /work2/noaa/hwrf/save/jcheng/GSI/GSItest/build Start 1: global_4denvar 1/7 Test #1 <#1>: global_4denvar ................... Passed 1326.62 sec Start 2: rtma 2/7 Test #2 <#2>: rtma ............................. Passed 965.50 sec Start 3: rrfs_3denvar_glbens 3/7 Test #3 <#3>: rrfs_3denvar_glbens ..............***Failed 425.44 sec Start 4: netcdf_fv3_regional 4/7 Test #4 <#4>: netcdf_fv3_regional ..............***Timeout 86400.06 sec Start 5: hafs_4denvar_glbens 5/7 Test #5 <#5>: hafs_4denvar_glbens ..............***Timeout 86400.04 sec Start 6: hafs_3denvar_hybens — Reply to this email directly, view it on GitHub <[#698 (comment)](https://github.com/NOAA-EMC/GSI/pull/698#issuecomment-1976497248)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADSYWF35GEJNWKKVDH6NWILYWRTWRAVCNFSM6AAAAABDJG6Q3GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZWGQ4TOMRUHA> . You are receiving this because you were assigned.Message ID: ***@***.***>

No I didn't. I will relaunch the regression test after modify it.

JingCheng-NOAA commented:
> Jing Did you do this as described in the previous comments “ To run regession tests for this PR, ush_cntrl as a env variable should be defined to point to the ush dir of the control GSI branch for different sub_hercules in ush dir are used for control and update runs).” Regard Ting
> […](#)
> On Mon, Mar 4, 2024 at 7:41 AM JingCheng-NOAA ***@***.***> wrote: @JingCheng-NOAA <https://github.com/JingCheng-NOAA> and @guoqing-noaa <https://github.com/guoqing-noaa> Could you please let me know whether you are available to review this PR? I've been doing the regression test on Hercules since Friday. But so far, it seems like something not setting up correctly. I am double checking it to see which part went wrong. Test project /work2/noaa/hwrf/save/jcheng/GSI/GSItest/build Start 1: global_4denvar 1/7 Test #1 <#1>: global_4denvar ................... Passed 1326.62 sec Start 2: rtma 2/7 Test #2 <#2>: rtma ............................. Passed 965.50 sec Start 3: rrfs_3denvar_glbens 3/7 Test #3 <#3>: rrfs_3denvar_glbens ..............***Failed 425.44 sec Start 4: netcdf_fv3_regional 4/7 Test #4 <#4>: netcdf_fv3_regional ..............***Timeout 86400.06 sec Start 5: hafs_4denvar_glbens 5/7 Test #5 <#5>: hafs_4denvar_glbens ..............***Timeout 86400.04 sec Start 6: hafs_3denvar_hybens — Reply to this email directly, view it on GitHub <[#698 (comment)](https://github.com/NOAA-EMC/GSI/pull/698#issuecomment-1976497248)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADSYWF35GEJNWKKVDH6NWILYWRTWRAVCNFSM6AAAAABDJG6Q3GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZWGQ4TOMRUHA> . You are receiving this because you were assigned.Message ID: ***@***.***>

No I didn't. I will relaunch the regression test with this suggestion then.

JingCheng-NOAA commented:
After took Ting's suggestion on defining the `ush_cntrl` in the script, the test case `rrfs_3denvar_glbens` passed with no issue. 
However the regression test of `netcdf_fv3_regional` is failed. The log file from `hiproc` run states there are errors related to NetCDF.
` ncdf error NetCDF: HDF error
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.`  The `loproc` run is completed with no error.

TingLei-NOAA commented:
@JingCheng-NOAA , thanks a lot for your testing of it.   If you didn't run ctest with -VV option, could you rerun "ctest -VV -R netcdf " >& ctest-netcdf.out .  And share that ctest-netcdf.out with me?

JingCheng-NOAA commented:
> @JingCheng-NOAA , thanks a lot for your testing of it. If you didn't run ctest with -VV option, could you rerun "ctest -VV -R netcdf " >& ctest-netcdf.out . And share that ctest-netcdf.out with me?

@TingLei-daprediction The log file is locate on Hercules: /work2/noaa/hwrf/save/jcheng/GSI/GSItest/build/ctest_netcdf.out

Please let me know if you cannot access it.

TingLei-daprediction commented:
@JingCheng-NOAA  Yes, I have no permission to that file :(.

JingCheng-NOAA commented:
> @JingCheng-NOAA Yes, I have no permission to that file :(.

@TingLei-daprediction Please give it a try now.

TingLei-daprediction committed:
put back disabled I_MPI_EXTRA_FILESYSTEM for regression tests on Hercules

TingLei-NOAA commented:
Thanks to @JingCheng-NOAA regression tests, which revealed the same "HDF error" issues with I_MPI_EXTRA_FILESYSTEM enabled occur for the loproc run. Now, there are 3 conclusions from my point of view.   1) Jing's results indicated the current PR doesn't give a solution (with stable performances) for the issue with I_MPI_EXTRA_FILESYSTEM enabled, though it greatly improve the performance with this option turned on .  
2) In addition to the optimized wind IO and upgrading for the IO for cold start files,  the memory related slurm parameters for regression tests on RDHPCS  are refined (Thanks to Raghu Reddy's help at RDHPCS help desk). Especially, current PR with the tuned slurm parameters resolved the unidentical issue found with hafs_3dvar  https://github.com/NOAA-EMC/GSI/issues/697. Confirmed that issue came from not-enough memory.      3) I think this PR could go ahead with its' changes to be incorporated into EMC GSI, while the solution of https://github.com/NOAA-EMC/GSI/issues/694, by which this PR was originally motivated could be further explored /investigated while this PR and related collaborative investigation make the "cultprit"  much clearer . Possible pathways will be further discussed on that GSI issue.

TingLei-daprediction committed:
Merge branch 'NOAA-EMC:develop' into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

ShunLiu-NOAA commented:
@DavidHuber-NOAA Ting suggests you review this PR because you worked on a similar issue recently.

DavidHuber-NOAA commented on regression/regression_param.sh:
```diff
@@ -300,10 +300,10 @@ if [[ "$machine" = "Hera" ]]; then
 elif [[ "$machine" = "Hercules" ]]; then
    export OMP_STACKSIZE=2048M
-   export APRUN="srun -n \$ntasks --cpus-per-task=\$threads"
+   export APRUN="srun  -n \$ntasks --mem=0  --cpus-per-task=\$threads"
```
Based on the document shared by Raghu, `--cpus-per-task` should probably be removed from these srun calls.  This is true for Jet as well.

TingLei-NOAA replied:
@DavidHuber-NOAA  .  That is a little intricate  part for me. Quote from Raghue Reddy: 
```

In your example, your tasks don't get to use all the memory if you are not using all the cores.  It is for those instances that having:

APRUN='srun -n $ntasks --mem=0 --cpus-per-task=$threads'

is needed and each task will be able to use all the available memory.
```
So, for my understanding, using --mem=0 in the srun command line could let all available memories are used.    Since cpus-per-task is not used in the sbatch pararmeter setup,  that is why it is used in the command line.   That will be great if further comfirmation/clarification  would be made.

DavidHuber-NOAA replied:
I take it back.  It looks like `--cpus-per-task` should stay.
Referring to
https://rdhpcs-common-docs.rdhpcs.noaa.gov/wiki/index.php?title=Running_and_Monitoring_Jobs#Specifying_a_processor_layout_for_your_job_(uniform_layout_example)
and
https://rdhpcs-common-docs.rdhpcs.noaa.gov/wiki/index.php?title=Running_and_Monitoring_Jobs#Setting_the_--cpus-per-task_parameter,
the `--cpus-per-task` parameter is required when submitting `OpenMP` threaded jobs, so we do need it.  However, by using it, `--exclusive` is ignored.  Instead, to state that we want to use all available memory on the node, `--mem=0` should be specified, which you have done.  So all we need to add for Jet is `--mem=0`.  It seems `--exclusive` should perhaps be removed since it is ignored, but I am OK with it staying as long as `--mem=0` remains.

DavidHuber-NOAA replied:
@TingLei-daprediction After a conversation with Raghu, I see now that my interpretation of the `--exclusive` and `--cpus-per-task` flags was incorrect.  All three flags, `--cpus-per-task`, `--exclusive`, and `--mem=0` should be used.  So the changes you have made here are correct.  Sorry for the confusion.
@RussTreadon-NOAA I don't have permission to resolve this conversation.  Could you do that for me?

DavidHuber-NOAA commented on src/gsi/gsi_rfv3io_mod.f90:
```diff
@@ -1312,6 +1312,7 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
       if(mype == 0) then
         if (allocated(fv3lam_io_dynmetvars2d_nouv)) &
           write(6,*)' fv3lam_io_dynmetvars2d_nouv is ',(trim(fv3lam_io_dynmetvars2d_nouv(i)), i=1,ndynvario2d)
+           call flush(6)
```
This debug flush should probably be removed.

DavidHuber-NOAA reviewed (changes requested):
A couple of suggestions regarding srun flags.  In addition, `--cpus-per-task` should probably be removed from sub_jet and replaced with `--exclusive`.  I'm happy to run the tests on Jet if you make this change.

DavidHuber-NOAA commented:
@TingLei-daprediction it looks like `gcc` is failing to compile due to undefined MPI symbols ([see the log](https://productionresultssa3.blob.core.windows.net/actions-results/d3affd32-c9b2-40ff-8472-cc1bbcdc52eb/workflow-job-run-e51c6a04-e80a-56b0-2050-253f87a07f68/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-06T14%3A46%3A36Z&sig=M%2FYzloKxHj7wfLU7xs5%2BLo8ZzGyhWDIuDFWI6RuOkUI%3D&sp=r&spr=https&sr=b&st=2024-03-06T14%3A36%3A31Z&sv=2021-12-02)).  Even though the GSI does not run with `gcc`, it should still compile.  Can you please address the errors?  You can test build on Hera with `COMPILER=gnu ./build.sh`.

TingLei-daprediction committed:
removed 'call fush(6)' in gsi_rfv3io_mod.f90 and a change in sub_jet

TingLei-daprediction committed:
fix for undefined mpi parameters complained by gcc

TingLei-daprediction committed:
Merge branch 'feature/fv3reg_parallel_io_upgrade' of https://github.com/TingLei-daprediction/GSI into feature/fv3reg_parallel_io_upgrade

TingLei-NOAA commented:
@DavidHuber-NOAA  the fix for the problem reported by GNU had been added.  
About use of --exclusive and --mem=0  , I have forwarded Raghu's detailed explanation in that ticket to you and Russ.
Hope that helps.

DavidHuber-NOAA commented:
@TingLei-daprediction I apologize for leaving my review hanging.  Could you sync your branch with develop?  After that, I will run tests on Jet and finish my review.

TingLei-NOAA commented:
@DavidHuber-NOAA  Thanks a lot.  I have done so.

DavidHuber-NOAA commented:
@TingLei-daprediction Running tests on Jet (kjet, CentOS), the `rrfs_3denvar_glbens_loproc`, `hafs_3denvar_hybens_loproc`, and `hafs_4denvar_glbens_loproc` tests all crashed with segmentation faults.  The RRFS crashed at

https://github.com/NOAA-EMC/GSI/blob/74c28e3a2054d775d7ca9378ad0d8c761ec35c3b/src/gsi/gsi_rfv3io_mod.f90#L4133

and the HAFS tests crashed one line later at

https://github.com/NOAA-EMC/GSI/blob/74c28e3a2054d775d7ca9378ad0d8c761ec35c3b/src/gsi/gsi_rfv3io_mod.f90#L4134

I then ran just the `rrfs_3denvar_glbens` test on Hera and encountered the same error at the same line.  The log file for this test is available here: `/scratch1/NCEPDEV/global/David.Huber/para/ptmp/test_pio/tmpreg_rrfs_3denvar_glbens/rrfs_3denvar_glbens_loproc_updat/stdout`.

DavidHuber-NOAA commented:
@TingLei-daprediction An update.  I reran the RRFS test on Jet after changing the job configuration:

```console
-           topts[1]="0:15:00" ; popts[1]="20/1/"  ; ropts[1]="/1"
-           topts[2]="0:15:00" ; popts[2]="20/2/"  ; ropts[2]="/1"
+           topts[1]="0:15:00" ; popts[1]="10/2/"  ; ropts[1]="/1"
+           topts[2]="0:15:00" ; popts[2]="10/4/"  ; ropts[2]="/1"
```

Perhaps these job configurations would work for the HAFS tests as well.  Hera and Jet-kjet have the same node configurations (40 cores, 96 GB RAM), so these job configurations should work for Hera as well.

TingLei-NOAA commented:
@DavidHuber-NOAA  . So many thanks.  Seems I made some mistakes on managing the changes on different machines. 
I had reported in https://github.com/NOAA-EMC/GSI/issues/697.  
```
An update on the "same" issue with rrfs_3denvar_glbens on hera with the branch feature/fv3reg_parallel_io_upgrade.

when "SBATCH --nodes=1 --ntasks-per-node=20 " , the rrfs_3denvar_glbens_loproc_updat would abort with error messages as : "Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))" and it indicated the error occurring with the following line in Bold:
    call check( nf90_var_par_access(gfile_loc, ugrd_VarId, nf90_collective))
    ...
**call check( nf90_get_var(gfile_loc,ugrd_VarId,work_bu,start=u_startloc,count=u_countloc) )**
If the previous line (set nf90_collective) was commented out, the GSI would run smoothly. But since that set up is needed for runs on hercules, It can't be commented out.
2) when "SBATCH --nodes=2 --ntasks-per-node=10 ", GSI run would run smoothly, but the issue (differences in fv3_dynvars between loproc_updat and hiproc_updat) occurred.
3) when "SBATCH --nodes=4 --ntasks-per-node=5 " , the issue disappeared.
In summary, seems in the "same"/similar issue with rrfs test on hera, it is related to the allocated memory resources for the low level parallel netcdf IO processes (especially when nf90_collective is used, larger memory usage would be caused). If there is no "enough" memory is given, the GSI could still run successfully, but possible memory corruption in the parallel netcdf io step could cause those differences investigated in this issue.
So, when we have been focusing on GSI codes and compiling packages and possible env variable, we might also test the impacts of the sbach setup especially helping allocating "enough" memory to the low-level netcdf/hdf5 IO.
```
So, the setup for those jobs should request larger memory than, even, the one you changed to.     I will update those opts for hera and jets .  And, I will let you know when they are cleaned .    My apologies.

TingLei-daprediction committed:
re-correction of a few slurm parameters for regression tests on hera and jet

DavidHuber-NOAA commented on regression/regression_param.sh:
```diff
@@ -87,17 +87,17 @@ case $regtest in
         elif [[ "$machine" = "Jet" ]]; then
-           topts[1]="0:15:00" ; popts[1]="20/1/"  ; ropts[1]="/1"
-           topts[2]="0:15:00" ; popts[2]="20/2/"  ; ropts[2]="/1"
+           topts[1]="0:15:00" ; popts[1]="5/5/"  ; ropts[1]="/1"
```
This should probably be `5/4` to match Hera:

```suggestion
           topts[1]="0:15:00" ; popts[1]="5/4/"  ; ropts[1]="/1"
```

TingLei-NOAA replied:
Done . Thanks

DavidHuber-NOAA commented on regression/regression_param.sh:
```diff
@@ -117,17 +117,17 @@ case $regtest in
         elif [[ "$machine" = "Orion" ]]; then
-           topts[1]="0:15:00" ; popts[1]="20/1/" ; ropts[1]="/1"
-           topts[2]="0:15:00" ; popts[2]="20/2/" ; ropts[2]="/2"
+           topts[1]="0:15:00" ; popts[1]="5/4/" ; ropts[1]="/1"
+           topts[2]="0:15:00" ; popts[2]="10/4/" ; ropts[2]="/2"
         elif [[ "$machine" = "Hercules" ]]; then
            topts[1]="0:15:00" ; popts[1]="20/1/" ; ropts[1]="/1"
-           topts[2]="0:15:00" ; popts[2]="20/2/" ; ropts[2]="/2"
+           topts[2]="0:15:00" ; popts[2]="5/8/" ; ropts[2]="/1"
```
Should the number of threads be 1 for all `hafs_3denvar_hybens` tests?  If so, then should Orion have its `hiproc` thread count reduced to match Hercules?

TingLei-NOAA replied:
@DavidHuber-NOAA , now, I changed those threads setup in this PR to the current head of EMC GSI has because I don't have any reasons to change them.    Since there are changes related to hercules too, I will re-run regression tests on hercules. Please let me know what you think about that.

TingLei-NOAA replied:
BTW:  the regression tests with the current popts setup for hera have all passed on hera (including that hafs_4densvar_glbens 's failing for "exceeded maximum allowable threshhold time.." was ignored)

TingLei-daprediction committed:
fix of potps on jet to match that on hera for rrfs_3denvar_glbens regression test as D Huber suggested

edwardhartnett commented:
All, for future reference, there is an internal logging capability in netcdf-c which might help with these kinds of problems. For parallel programs, the netcdf-c library generates a log file for each processor, with detailed information about what netCDF functions are called. Logging needs to be turned on in netcdf-c for this to work, but we can arrange that.

It's documented here: https://github.com/Unidata/netcdf-c/blob/main/docs/logging.md

TingLei-daprediction committed:
fix of potps on jet to match that on hera for rrfs_3denvar_glbens regression test as D Huber suggested

TingLei-daprediction committed:
Merge branch 'NOAA-EMC:develop' into feature/fv3reg_parallel_io_upgrade

TingLei-daprediction committed:
unanimously assign 4 nodes for regression tests on RDHPCS machines

TingLei-NOAA commented:
All regression tests passed on orion while for rrfs_3denvar_glbens with the "Failure time-thresh"  ignored. 
On hercules, only hafs_4densvar_glbens failed for the control runs (both loproc and hiproc) produced different fv3_dynvars as investigated in https://github.com/NOAA-EMC/GSI/issues/697.  So, I would suggest this test failure could be ignored regarding this PR.

DavidHuber-NOAA commented:
@TingLei-daprediction Thanks for updating the job CPU/node counts.  Can you sync your branch with develop?  Once that is done, I will restart testing on Jet.

TingLei-daprediction committed:
Merge branch 'NOAA-EMC:develop' into feature/fv3reg_parallel_io_upgrade

TingLei-NOAA commented:
@DavidHuber-NOAA .  Thanks!  "Sync " has been done.

DavidHuber-NOAA reviewed (approved):
Changes look good and regression tests pass on Jet.  Approve.

ShunLiu-NOAA commented:
@JingCheng-NOAA and @XuLu-NOAA If you are available, please review Ting's PR again. Thank you.

JingCheng-NOAA commented:
I've tested Ting's latest update again on Hercules. The `hafs_4densvar_glbens` failed as expected, which should not affect this PR. But `global_4denvar` also failed, which is due to access issue to the prepbufr file `/work/noaa/da/rtreadon/CASES/regtest/gfs/prod/gdas.20240223/00/obs/gdas.t00z.prepbufr` that used by this case.
All other test cases passed.

RussTreadon-NOAA commented:
`gdas.t00z.prepbufr` is a restricted access file.  Users must belong to the `rstprod` group to access this file.  The following files in `/work/noaa/da/rtreadon/CASES/regtest/gfs/prod/gdas.20240223/00/obs/` are `rstprod` files
```
-rw-r----- 1 rtreadon rstprod   18764472 Feb 23 05:51 gdas.t00z.adpsfc.tm00.bufr_d
-rw-r----- 1 rtreadon rstprod   19164416 Feb 23 05:50 gdas.t00z.aircar.tm00.bufr_d
-rw-r----- 1 rtreadon rstprod    3667912 Feb 23 05:50 gdas.t00z.aircft.tm00.bufr_d
-rw-r----- 1 rtreadon rstprod     287840 Feb 23 05:50 gdas.t00z.gpsipw.tm00.bufr_d
-rw-r----- 1 rtreadon rstprod   44055880 Feb 23 05:52 gdas.t00z.gpsro.tm00.bufr_d
-rw-r----- 1 rtreadon rstprod   11669280 Feb 23 06:02 gdas.t00z.nsstbufr
-rw-r----- 1 rtreadon rstprod  101586864 Feb 23 06:02 gdas.t00z.prepbufr
-rw-r----- 1 rtreadon rstprod   16949456 Feb 23 06:02 gdas.t00z.prepbufr.acft_profiles
-rw-r----- 1 rtreadon rstprod          0 Feb 23 05:53 gdas.t00z.saphir.tm00.bufr_d
-rw-r----- 1 rtreadon rstprod    9692304 Feb 23 05:50 gdas.t00z.sfcshp.tm00.bufr_d
```

JingCheng-NOAA commented:
> > @JingCheng-NOAA and @guoqing-noaa Could you please let me know whether you are available to review this PR?
> 
> I've been doing the regression test on Hercules since Friday. But so far, it seems like something not setting up correctly. I am double checking it to see which part went wrong. Test project /work2/noaa/hwrf/save/jcheng/GSI/GSItest/build Start 1: global_4denvar 1/7 Test #1: global_4denvar ................... Passed 1326.62 sec Start 2: rtma 2/7 Test #2: rtma ............................. Passed 965.50 sec Start 3: rrfs_3denvar_glbens 3/7 Test #3: rrfs_3denvar_glbens ..............***Failed 425.44 sec Start 4: netcdf_fv3_regional 4/7 Test #4: netcdf_fv3_regional ..............***Timeout 86400.06 sec Start 5: hafs_4denvar_glbens 5/7 Test #5: hafs_4denvar_glbens ..............***Timeout 86400.04 sec Start 6: hafs_3denvar_hybens

Then has the link of the data of prepbufr in regression test changed? Because two weeks ago, the `global_4denvar` case passed in my tests on Hercules.
Other than this issue, I have no more questions about this PR.

RussTreadon-NOAA commented:
The case for the global ctests was recently updated to bring in GMI data.  The previous global case did **NOT** properly restrict `rstprod` data.

ShunLiu-NOAA commented:
@JingCheng-NOAA Thank you Jing for the test. Since Ting already completed regression test on WCOSS2 and Orion, @DavidHuber-NOAA completed the test on JET. We may consider merging this PR to develop.

TingLei-NOAA commented:
@DavidHuber-NOAA  @XuLu-NOAA @JingCheng-NOAA  , Thanks for your help as the reviewers.

TingLei-daprediction commented:
@ShunLiu-NOAA  Thanks!
